src/Razor/src/rzls/Program.cs

```diff
@@ -3,33 +3,39 @@
 
 using System;
 using System.Diagnostics;
+using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer;
 
 public class Program
 {
-    public static void Main(string[] args)
-    {
-#pragma warning disable VSTHRD002 // Avoid problematic synchronous waits
-        MainAsync(args).Wait();
-#pragma warning restore VSTHRD002 // Avoid problematic synchronous waits
-    }
-
-    public static async Task MainAsync(string[] args)
+    public static async Task Main(string[] args)
     {
         var trace = Trace.Messages;
+
         for (var i = 0; i < args.Length; i++)
         {
             if (args[i].Contains("debug", StringComparison.OrdinalIgnoreCase))
             {
-                while (!Debugger.IsAttached)
+                await Console.Error.WriteLineAsync($"Server started with process ID {Environment.ProcessId}");
+                if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                {
+                    // Debugger.Launch() only works on Windows.
+                    _ = Debugger.Launch();
+                }
+                else
                 {
-                    Thread.Sleep(1000);
+                    var timeout = TimeSpan.FromMinutes(1);
+                    await Console.Error.WriteLineAsync($"Waiting {timeout:g} for a debugger to attach");
+                    using var timeoutSource = new CancellationTokenSource(timeout);
+                    while (!Debugger.IsAttached && !timeoutSource.Token.IsCancellationRequested)
+                    {
+                        await Task.Delay(100, CancellationToken.None);
```

Comment:
Did you mean to pass `timeoutSource.Token` here?

Reply:
I'm going to plead the 5th: I just copied David B's code, and I don't think it matters much for this scenario anyway